src/libraries/System.Private.CoreLib/src/System/Convert.cs

```diff
@@ -235,49 +193,49 @@ internal static object DefaultToType(IConvertible value, Type targetType, IForma
             }
 
-            if (ReferenceEquals(value.GetType(), targetType))
+            if (value.GetType() == targetType)
             {
                 return value;
             }
 
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.Boolean]))
+            if (targetType == typeof(bool))
                 return value.ToBoolean(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.Char]))
+            if (targetType == typeof(char))
                 return value.ToChar(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.SByte]))
+            if (targetType == typeof(sbyte))
                 return value.ToSByte(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.Byte]))
+            if (targetType == typeof(byte))
                 return value.ToByte(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.Int16]))
+            if (targetType == typeof(short))
                 return value.ToInt16(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.UInt16]))
+            if (targetType == typeof(ushort))
                 return value.ToUInt16(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.Int32]))
+            if (targetType == typeof(int))
                 return value.ToInt32(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.UInt32]))
+            if (targetType == typeof(uint))
                 return value.ToUInt32(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.Int64]))
+            if (targetType == typeof(long))
                 return value.ToInt64(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.UInt64]))
+            if (targetType == typeof(ulong))
                 return value.ToUInt64(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.Single]))
+            if (targetType == typeof(float))
                 return value.ToSingle(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.Double]))
+            if (targetType == typeof(double))
                 return value.ToDouble(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.Decimal]))
+            if (targetType == typeof(decimal))
                 return value.ToDecimal(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.DateTime]))
+            if (targetType == typeof(DateTime))
                 return value.ToDateTime(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.String]))
+            if (targetType == typeof(string))
                 return value.ToString(provider);
-            if (ReferenceEquals(targetType, ConvertTypes[(int)TypeCode.Object]))
+            if (targetType == typeof(object))
```

Comment:
This is going to introduce performance regression:

```
using System;

class Test
{
    static void Main()
    {
        var o = (int)42; 
        int start = Environment.TickCount;
        for (int i = 0; i < 100000000; i++)
            Convert.ChangeType(o, typeof(double), null);
        int end = Environment.TickCount;
        Internal.Console.WriteLine((end - start).ToString());   
    }
}
```

Baseline: 2.2s With this change: 4.1s

Reply:
I am wondering whether a better strategy would be to make `typeof(...)` faster by compiling it into de-reference of handle holding the object. The handle can be initialized lazily, similar to how we initialize dictionary entries lazily. 

E.g. something like this:

```
mov rax, handle address
mov rax, [rax]
test  rax, rax
jne Done
mov rcx, MethodTable*
call CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE
Done

Reply:
Well yeah, instead of comparing pointers, we now have to go to through a helper that is fast, but not as fast as comparing two pointers. It's a startup optimization, really. It would probably be within noise range had the cascading `if` in `Convert` been shorter, but it's not.

Redoing this into into handle is more work than I would be willing to sign up for.